### PR TITLE
Authentication Refactor

### PR DESCRIPTION
**Pull request by texm**
On the frontend:
- `auth` changed to a module, with functions for logging in and out (using firebase for account retrieval, backend for auth/session creation)
  - Still needs work to persist the auth state across refreshes
- AuthContext created which all components under app can use, holds `user` object and current login state
- Data fetching moved from app to a child component `Content`, can use AuthContext
- We still need to stop using `user.id` in calls to the backend in favour of storing the id in the session

On the backend:
- TestSessionCookieFilter created as an example of using filters. It checks for a valid Firebase session cookie, and logs the users email.
  - We should adapt this logic into an `AuthenticationFilter` that checks the session cookie for a valid Firebase session cookie AND a valid user id in the regular session. A stub for setting the user id in the session is line 106 in AuthServlet.java
  - This filter is matched on any url that starts with `/api`
  - TestApiServlet created as an example, with calls made from the frontend to demonstrate. It will respond 200 if the user is authenticated

You can actually compare changes to master here: [https://github.com/googleinterns/step272-2020/compare/master...fe-auth-fix](https://github.com/googleinterns/step272-2020/compare/master...fe-auth-fix) (github does not show changes to target branch)

### CODE REVIEW

File: project/src/main/java/com/google/sps/servlets/AuthServlet.java [hunk 1]
```diff
@@ -89,17 +75,75 @@ private String verifyIDToken(HttpServletRequest request) throws IOException{
...
+  }
+
+  /**
+   * If user is in the database, get the user ID in the database from the given email.
+   * Otherwise add user to the database and return the user's ID.
+   * @param request -- HTTP request
+   * @param email -- user's email from the decoded token
+   * @return -- user's ID in the database
+   */
+  private long getUserID(HttpServletRequest request, String email){
```
Comment (texm) [inline project/src/main/java/com/google/sps/servlets/AuthServlet.java:130]: I think this function name should convey that a user will also be created if not found in the database - eg `getOrCreateUserID`

File: project/frontend/src/services/auth.js [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+          .then((data) => {
+            callback(true, {
+              email: user.email,
+              displayName: user.displayName,
+              id: data.id,
+            });
+            console.log("user id: " + data.id);
+          });
+      });
+    });
```
Comment (texm) [inline project/frontend/src/services/auth.js:38]: we should add a `catch` here to feed a `false` to the `callback` function if anything in the flow fails @Phoebe-Khokgawe

Comment (Phoebe-Khokgawe) [reply project/frontend/src/services/auth.js:38]: yup

Comment (Phoebe-Khokgawe) [reply project/frontend/src/services/auth.js:42]: Added here 09661af7fbdaf46c72494d7deecf4fcdba1f89c1
Will it be better to add it to all the promises or just the final one here is enough.

**Comment by Phoebe-Khokgawe** [comment]
@googlebot I consent.

**Review by leonnemets — APPROVED** [review]

**Comment by lin8231** [comment]
* Persisted state of user (store/access/delete via storage)
* Fixed bug where content would not rerender when compId changes
* Fix a merge issue where a chunk of code got deleted

**Review by lin8231 — APPROVED** [review]

**Pull request merged**